lib/commander/user_interaction.rb

```diff
@@ -24,6 +24,44 @@ module UI
       include Growl
     end
 
+    #--
+    # Mimic growl in linux with libnotify
+    #++
+
+    begin
+      require 'libnotify'
+    rescue LoadError
+      # Do nothing
+    else
+      Libnotify.icon_dirs << '/usr/share/icons/gnome/*/'
+      module GrowlLibnotify
+        # Growl offers a simple api for macos desktop notifications.
+        # It is included by default by commander.
+        #
+        # This mimics that api using libnotify for linux desktop notifications.
+
+        # https://github.com/splattael/libnotify
+
+        {
+          nil => nil,
+          info: 'info',
+          ok: 'sunny',
+          warning: 'important',
+          error: 'error',
+        }.each_pair do |status, icon|
+          define_method ['notify', status].compact.join('_') do |message, *_args|
+            Libnotify.new do |notify|
+              notify.summary = message unless status.nil?
+              notify.body = message
+              notify.transient = true
+              notify.icon_path = icon
+            end.show!
+          end
+        end
+      end
```

Comment:
I would ideally like to see some test coverage of this code, probably stubbing libnotify itself but at least testing that the invocations are correct for each method.

Reply:
A little tricky with the current organization. Perhaps, if you don't mind, I could extract the GowlLibnotify into its own file (growl_libnotify.rb) alongside user_interaction.rb. Easy enough to mock/stub libnotify and spec out that the methods get defined anyhow -- or I could unroll them into their five individual definitions again (removing all meta programing). Beyond that though, if they are defined then the invocation is just method_name(str), not much to test there without libnotify itself.

Reply:
I'm fine with extracting the module. What I would test is that a stubbed `Libnotify` module receives the messages you expect when your module's `notify`, `notify_ok`, etc. methods are called.